### PR TITLE
TCP: per-client idle timeout for tcp conn pool

### PR DESCRIPTION
**Pull request by JuniorHsu**
Signed-off-by: kuochunghsu <kuochunghsu@pinterest.com>

Commit Message:

While we return the connection to the pool, the idle timeout starts ticking until the connection is reused. If the timeout fires, we closes the connection and let the pool knows.

Additional Description:
Risk Level: mid
Testing: unit test, had a e2e test
[Optional Fixes #Issue] #23699

### CODE REVIEW

**Comment by JuniorHsu** [comment]
This is WIP as the unit test is not done yet.
Default idle timeout is 1 hour.
We did an e2e test on a ~40 QPS downstream host.
After running 5 hours, the memory usage is 4% less than the host without idle timeout.
The connection reuse rate  `1 - (cx_total/rq_total)` is about the same (99.7%)

**Comment by JuniorHsu** [comment]
cc @alyssawilk @zuercher @rgs1 @fishcakez @PeterL328

File: api/envoy/config/cluster/v3/cluster.proto
Comment (alyssawilk) [inline api/envoy/config/cluster/v3/cluster.proto:1192]: Is this then on by default?  If so we should comment on the default timeout and runtime guard the behavioral change.

File: api/envoy/config/cluster/v3/cluster.proto
Comment (alyssawilk) [inline api/envoy/config/cluster/v3/cluster.proto:1194]: If this is TCP only I think it should be in tcp-specific config, kind of like the upstream http protocol options.
I think it makes sense as a general field, though we'd need to put some thought into how it interacts with prefetching connections.

File: envoy/upstream/upstream.h
Comment (alyssawilk) [inline envoy/upstream/upstream.h:814]: how does tihs differ from idleTimeout?  At a glance it looks to me like it's possible all we need is a tcp options parallel to the http protocol options so we can set idleTimeout for non-tcp-upstreams. WDIT?

**Review by alyssawilk — COMMENTED** [review]
thanks for tackling this!
A couple of high level thoughts before I dive into details
/wait

**Comment by alyssawilk** [comment]
FWIW I'm fine with you de-drafting this and just putting NEEDED in the test field.  We have slack machinery for reminding us to do reviews and I might otherwise lost track of it :-)

File: api/envoy/config/cluster/v3/cluster.proto
Comment (JuniorHsu) [reply api/envoy/config/cluster/v3/cluster.proto:1192]: Yes I'll add a runtime guard. Thanks for pointing this out.

File: api/envoy/config/cluster/v3/cluster.proto
Comment (JuniorHsu) [reply api/envoy/config/cluster/v3/cluster.proto:1194]: tcp pool is used by several L7 protocol like thrift so I'm with you to put as a general field. 

Default idle timeout (with runtime guard unblock) would be an hour, and the timer starts ticking only if no one holds the connection. Does prefetching in this context mean creating at most 3 more connections for every new connection? If yes, I wouldn't worry about time performance for  1 hour idle timeout. That hurts only if we don't need that much connection every moment in previous hour.

File: envoy/upstream/upstream.h
Comment (JuniorHsu) [reply envoy/upstream/upstream.h:814]: I assumed you were saying `non-http-upstreams`

`idleTimeout` here is the same idea with `tcpPoolIdleTimeout` so we can use `idleTimeout` for both http codec and tcp pool.

`idleTimeout` could be what http protocol options say or what cluster proto says, as we need separated configs. Does it make sense?

**Comment by JuniorHsu** [comment]
> FWIW I'm fine with you de-drafting this and just putting NEEDED in the test field. We have slack machinery for reminding us to do reviews and I might otherwise lost track of it :-)

Thanks for your great input! I'm going to work on unit test as I put NEEDED 😅

**Comment by JuniorHsu** [comment]
On the other hand, in our test environment, 1 hour tcp idle timeout save 10% envoy memory after running 3 days
cx_total is pretty much the same 2316 vs 2294 for one upstream and didn't observe significant difference with p99 round trip time

File: envoy/upstream/upstream.h
Comment (JuniorHsu) [reply envoy/upstream/upstream.h:861]: I eventually keep it separated since i'm not sure if we can have a cluster has both idleTimeout for http and tcpPoolIdleTimeout for tcp pool. happy to make it to one if it's not possible to have both

File: api/envoy/config/cluster/v3/cluster.proto
Comment (mattklein123) [inline api/envoy/config/cluster/v3/cluster.proto:1193]: Is there a default if not set? If so how does one disable? Can we document that?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Small API question, thanks.

File: api/envoy/config/cluster/v3/cluster.proto
Comment (JuniorHsu) [reply api/envoy/config/cluster/v3/cluster.proto:1193]: great point. the is guard by reloadable flags. If reloadable is on, default is an hour, and setting 0 to disable it. I'll document it. Thanks

**Review by mattklein123 — APPROVED** [review]
API LGTM, thanks.

File: api/envoy/config/cluster/v3/cluster.proto
Comment (alyssawilk) [reply api/envoy/config/cluster/v3/cluster.proto:1200]: sorry @mattklein123 can I get your thoughts on this comment?
We have an HTTP equivalent and it's in the http-specific upstream config.  If we're going to leave this here I think we either have to deprecate the HTTP specific option, or at least mention the fact this field is ignored for HTTP upstreams.

Comment (mattklein123) [reply api/envoy/config/cluster/v3/cluster.proto:1200]: Oh I see sorry I missed this. Yeah can we just have a TCP protocol specific options much like we do for HTTP?

File: api/envoy/config/cluster/v3/cluster.proto
Comment (alyssawilk) [inline api/envoy/config/cluster/v3/cluster.proto:1189]: sorry I think I wasn't clear on how this works for HTTP :-/  For HTTP, we have this:

https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/upstreams/http/v3/http_protocol_options.proto

the deal is that you can stuff largely any proto in typed_extension_protocol_options  here:
https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/cluster/v3/cluster.proto#L937

so don't need to declare them in the cluster proto, and folks not needing that proto type don't need to pull it in.  So I think we want to treat TCP as we do HTTP - have a standalone proto (which can live in its own file) and it's added via that type extension config field.  Make sense?

/wait

Comment (JuniorHsu) [reply api/envoy/config/cluster/v3/cluster.proto:1189]: @alyssawilk oh thanks for letting me know. It takes some time to make me aware most of the protocol_options field in cluster.proto are deprecated
https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto

HTTP proto structure is like
```
api/envoy/extensions/upstreams/http $ tree .
.
├── generic
│   └── v3
│       ├── BUILD
│       └── generic_connection_pool.proto
├── http
│   └── v3
│       ├── BUILD
│       └── http_connection_pool.proto
├── tcp
│   └── v3
│       ├── BUILD
│       └── tcp_connection_pool.proto
└── v3
    ├── BUILD
    └── http_protocol_options.proto
```

current tcp is 
```
api/envoy/extensions/upstreams/tcp $ tree
.
└── generic
    └── v3
        ├── BUILD
        └── generic_connection_pool.proto
```

So i'm going to add a v3 folder at the root of tcp and put the idle timeout into new standalone tcp_protocol_options.proto. the sturcture will be 
```
api/envoy/extensions/upstreams/tcp $ tree
.
├── generic
│   └── v3
│       ├── BUILD
│       └── generic_connection_pool.proto
└── v3
    ├── BUILD
    └── tcp_protocol_options.proto
```

sounds like a plan? @alyssawilk

Comment (alyssawilk) [reply api/envoy/config/cluster/v3/cluster.proto:1189]: I'm going to let @mattklein123 as I think it's more of an API shephard call :-)
cc @envoyproxy/api-shepherds

Comment (mattklein123) [reply api/envoy/config/cluster/v3/cluster.proto:1189]: Yup this sounds good to me. Thank you.

Comment (JuniorHsu) [reply api/envoy/config/cluster/v3/cluster.proto]: Updated. fwiw i hit
` This change appears to add visibility rules. Please get senior maintainer approval to add an exemption to check_visibility tools/code_format/check_format.py` since I make the new proto public visibility

File: source/extensions/upstreams/tcp/BUILD
Comment (JuniorHsu) [inline source/extensions/upstreams/tcp/BUILD:16]: This triggers `This change appears to add visibility rules. Please get senior maintainer approval to add an exemption to check_visibility tools/code_format/check_format.py` and blocks ci

Could you help to move forward? @mattklein123 @alyssawilk 
Thanks

Comment (mattklein123) [reply source/extensions/upstreams/tcp/BUILD:16]: I think this is fine as it maps to what we do for HTTP but I will defer to Alyssa who did the HTTP work related to this.

**Review by alyssawilk — COMMENTED** [review]
Looking good!  Please check out the format issue on CI as well
/wait

File: api/envoy/extensions/upstreams/tcp/v3/tcp_protocol_options.proto
Comment (JuniorHsu) [reply api/envoy/extensions/upstreams/tcp/v3/tcp_protocol_options.proto:47]: I saw default is 1 hour everywhere else but happy to make it 10 minutes

File: source/extensions/upstreams/tcp/BUILD
Comment (JuniorHsu) [reply source/extensions/upstreams/tcp/BUILD:16]: Alyssa agrees on slack so I'm going to tweek `tools/code_format/config.yaml`

File: api/envoy/extensions/upstreams/tcp/v3/tcp_protocol_options.proto
Comment (alyssawilk) [inline api/envoy/extensions/upstreams/tcp/v3/tcp_protocol_options.proto:35]: for connection -> for the connection

File: api/envoy/extensions/upstreams/tcp/v3/tcp_protocol_options.proto
Comment (alyssawilk) [inline api/envoy/extensions/upstreams/tcp/v3/tcp_protocol_options.proto:36]: defined as the period in which there connection is not associated with a downstream connection

File: api/envoy/extensions/upstreams/tcp/v3/tcp_protocol_options.proto
Comment (alyssawilk) [inline api/envoy/extensions/upstreams/tcp/v3/tcp_protocol_options.proto:38]: I think we can skip this - it's in the release notes and is envoy-specific

File: docs/root/faq/configuration/timeouts.rst
Comment (alyssawilk) [inline docs/root/faq/configuration/timeouts.rst:161]: again if the connection is not in use

File: source/common/upstream/upstream_impl.cc
Comment (alyssawilk) [inline source/common/upstream/upstream_impl.cc:1111]: doesn't applying the runtime guard here mean it won't actually reload unless you reload the cluster?  I think we may want to instead create the idle alarm based on the runtime guard such that the flag takes effect on connection creation

File: test/common/tcp/conn_pool_test.cc
Comment (alyssawilk) [inline test/common/tcp/conn_pool_test.cc:424]: Is it possible to get an e2e test as well?  If there's no existing e2e tests with idle connection reuse we could configure it so we reclaim eagerly (1s or less), turn up connection preconnect so that a downstream request kicks off 2 upstream connections and then wait on the stat for the upstream connection to be reclaimed?

**Review by alyssawilk — COMMENTED** [review]
Awesome.  A few more thoughts (sorry for not picking them up last round)

File: source/common/upstream/upstream_impl.cc
Comment (JuniorHsu) [reply source/common/upstream/upstream_impl.cc:1111]: You mean we'll reload the cluster only if two `ClusterInfoImpl` objects around CDS/guard_flip are not equal? Yeah happy to defer the timer decision to connection creation.

File: test/common/tcp/conn_pool_test.cc
Comment (JuniorHsu) [reply test/common/tcp/conn_pool_test.cc:424]: Does e2e in the context mean integration test? We did test in our environment for days.

Comment (alyssawilk) [reply test/common/tcp/conn_pool_test.cc:426]: e2e means /test/integration/tcp_proxy_integration_test.cc :-)  I'm glad you tested internally (that's actually pretty awesome) but it'd be good to have it also end to end regression tested.
I think if you set your new configuration option to a short timeout you can adapt one of the tests which does Preconnect and verify the connection gets reclaimed by waiting on cluster.cluster_0.upstream_cx_destroy

**Comment by alyssawilk** [comment]
@ggreenway this looks good modulo the missing integration test.  Would you be game for second pass?

File: api/envoy/extensions/upstreams/tcp/v3/tcp_protocol_options.proto
Comment (ggreenway) [inline api/envoy/extensions/upstreams/tcp/v3/tcp_protocol_options.proto:36]:
```suggestion
  // the connection is not associated with a downstream connection. When the idle timeout is
```

File: changelogs/current.yaml
Comment (ggreenway) [inline changelogs/current.yaml:230]: I think this should also be mentioned in section `minor_behavior_changes` because we're adding a timeout that's configured by default, where there wasn't one before.

File: docs/root/faq/configuration/timeouts.rst
Comment (ggreenway) [inline docs/root/faq/configuration/timeouts.rst:161]: `the connection is not in use`. I think it would be helpful to define here (as you did in other places in the docs) exactly what that means, as one reasonable interpretation of `not in use` is `no data being transferred in either direction`, but that is a covered by the tcpproxy idle timeout, not this one.

File: source/common/tcp/conn_pool.cc
Comment (ggreenway) [inline source/common/tcp/conn_pool.cc:66]: Can the previous line result in this connection getting closed? If so, maybe move this line up.

File: source/common/tcp/conn_pool.h
Comment (ggreenway) [inline source/common/tcp/conn_pool.h:129]:
```suggestion
  // Called if the underlying connection is idle over the cluster's tcpPoolIdleTimeout()
```

**Review by ggreenway — CHANGES_REQUESTED** [review]
Looks good overall

/wait

File: source/common/tcp/conn_pool.h
Comment (JuniorHsu) [reply source/common/tcp/conn_pool.h]: :facepalm: thanks for catching this...

File: source/common/tcp/conn_pool.cc
Comment (JuniorHsu) [reply source/common/tcp/conn_pool.cc:65]: Nice catch. Thanks!

File: changelogs/current.yaml
Comment (JuniorHsu) [reply changelogs/current.yaml]: I move to `minor_behavior_changes` and put a see also here in new feature

**Review by alyssawilk — APPROVED** [review]

File: test/integration/tcp_proxy_integration_test.cc
Comment (ggreenway) [inline test/integration/tcp_proxy_integration_test.cc:686]: What was the failure? I'd like to have some understanding of it before disabling a test on a platform.

**Review by ggreenway — CHANGES_REQUESTED** [review]
/wait-any

File: test/integration/tcp_proxy_integration_test.cc
Comment (JuniorHsu) [reply test/integration/tcp_proxy_integration_test.cc:686]: I don't have apple build on hand so here's what I can tell.

The log simply shows timeout for apple platform but that might be testing environment issue.
While I wrote the e2e test without preconnections, 
I grab a connection from pool, close the connection to return it to pool, the idle timer starts ticking but never fire alarms. 
I was hinted to use preconnections, and those timers fire alarms as expected.

Not a thread issue since everything seems on the same thread
```
[ RUN      ] IpVersions/TcpProxyIntegrationTest.TestPerClientIdletimeout/IPv4
[2022-12-15 17:33:08.949][32224][error][pool] [source/common/tcp/conn_pool.cc:116] [C4] setIdleTimer
[2022-12-15 17:33:08.949][32224][error][pool] [source/common/tcp/conn_pool.cc:116] [C5] setIdleTimer
[2022-12-15 17:33:08.949][32224][error][pool] [source/common/tcp/conn_pool.cc:116] [C7] setIdleTimer
[2022-12-15 17:33:08.949][32224][error][pool] [source/common/tcp/conn_pool.cc:109] [C7] disableIdleTimer
[2022-12-15 17:33:08.950][32224][error][pool] [source/common/tcp/conn_pool.cc:109] [C7] disableIdleTimer
[2022-12-15 17:33:08.950][32224][error][pool] [source/common/tcp/conn_pool.cc:116] [C7] setIdleTimer
[2022-12-15 17:33:09.050][32224][error][pool] [source/common/tcp/conn_pool.cc:102] [C4] per client idle timeout
[2022-12-15 17:33:09.050][32224][error][pool] [source/common/tcp/conn_pool.cc:109] [C4] disableIdleTimer
[2022-12-15 17:33:09.050][32224][error][pool] [source/common/tcp/conn_pool.cc:102] [C5] per client idle timeout
[2022-12-15 17:33:09.050][32224][error][pool] [source/common/tcp/conn_pool.cc:109] [C5] disableIdleTimer
```

Hence I have a hunch there's a profound testing glitch in specific platform which is beyond my knowledge.
If this is a concern, I am happy to make idle time out happen without specific platform.

Comment (ggreenway) [reply test/integration/tcp_proxy_integration_test.cc:686]: Are you able to determine for sure whether it's an issue with the test, or an issue with the feature?

I tried running 1000 iterations of the test on my mac and saw no failures. If we can't reproduce it locally, this could be a challenge to figure out.

Comment (JuniorHsu) [reply test/integration/tcp_proxy_integration_test.cc:686]: It's a testing issue since I could reproduce locally (finger crossed) and `TCPActiveClient` is never created unless I put `
  test_server_->waitForCounterGe("cluster.cluster_0.upstream_cx_total", 1);` to wait. PR is coming

Comment (JuniorHsu) [reply test/integration/tcp_proxy_integration_test.cc]: I `git push` the update. Let's see if ci is happy

Comment (JuniorHsu) [reply test/integration/tcp_proxy_integration_test.cc]: Nice. Both fails last time.
https://dev.azure.com/cncf/envoy/_build/results?buildId=123667&view=logs&j=fa3d3e18-6969-5713-c3e7-3581195704fd&t=ba370bef-ca49-52f5-0c93-7c0c0f27c465&l=38777

```
[ RUN      ] IpVersions/TcpProxyIntegrationTest.TestPerClientIdletimeout/IPv4
[external/com_google_absl/absl/flags/internal/flag.cc : 115] RAW: Restore saved value of envoy_reloadable_features_no_extension_lookup_by_name to: true
[external/com_google_absl/absl/flags/internal/flag.cc : 115] RAW: Restore saved value of envoy_reloadable_features_runtime_initialized to: false
[       OK ] IpVersions/TcpProxyIntegrationTest.TestPerClientIdletimeout/IPv4 (398 ms)
...
[ RUN      ] IpVersions/TcpProxyIntegrationTest.TestPerClientIdletimeout/IPv6
[external/com_google_absl/absl/flags/internal/flag.cc : 115] RAW: Restore saved value of envoy_reloadable_features_runtime_initialized to: false
[external/com_google_absl/absl/flags/internal/flag.cc : 115] RAW: Restore saved value of envoy_reloadable_features_no_extension_lookup_by_name to: true
[       OK ] IpVersions/TcpProxyIntegrationTest.TestPerClientIdletimeout/IPv6 (383 ms)

```

**Comment by JuniorHsu** [comment]
@ggreenway tests pass, might to take another look? Thanks.

**Pull request merged**
